### PR TITLE
[Bug] Fix Jaw Lock leaving the user trapped after the target faints

### LINKED ISSUE #170: Killing a wild Pokémon with Jaw lock still leaves the user trapped in the next battle
What the title says. I haven't used Jaw Lock again to further investigate because it's too expensive to watch Dreadnaw be chipped away to death lol. I figure this only occurs during waves of wild Pokémon and should be working properly during trainer battles.

### PR DESCRIPTION
**Pull request by innerthunder**
## What are the changes?

This fixes an issue where if a Pokemon uses Jaw Lock on a target, then KOs the target on a following turn, the user would remain trapped.

## Why am I doing these changes?

This is a mirror of #2702 , which closes #170 

## What did change?

- `data/battler-tags`: New `JawLockTag` to implement Jaw Lock's trapping effect on a Pokemon
- `data/move`: Jaw Lock now uses a new attribute `AddLinkedBattlerTagAttr` that adds a tag to the user and target, with each tag's `sourceId` being of the other Pokemon.
- `test/moves/jaw_lock.test`: New integration test suite checking how Jaw Lock's battler tags are maintained in different battle states.

### Screenshots/Videos

See #2702 for test videos.

## How to test the changes?

`npm run test jaw_lock`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [ ] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/data/move.ts [hunk 1]
```diff
@@ -4390,6 +4390,30 @@ export class GulpMissileTagAttr extends MoveEffectAttr {
...
+ * Attribute used for moves that add a tag to the user and target, such as Jaw Lock
+ * @extends AddBattlerTagAttr
+ */
+export class AddLinkedBattlerTagAttr extends AddBattlerTagAttr {
+  apply(user: Pokemon, target: Pokemon, move: Move, args: any[]): boolean {
+    if (!super.canApply(user, target, move, args)) {
+      return false;
+    }
+
+    const moveChance = this.getMoveChance(user,target,move,this.selfTarget);
```
Comment (torranx) [inline src/data/move.ts:4403]:
```suggestion
    const moveChance = this.getMoveChance(user, target, move, this.selfTarget);
```

File: src/test/moves/jaw_lock.test.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+      const leadPokemon = game.scene.getPlayerPokemon()!;
+      const enemyPokemon = game.scene.getEnemyPokemon()!;
+
+      game.doAttack(getMovePosition(game.scene, 0, Moves.JAW_LOCK));
+
+      await game.phaseInterceptor.to(MoveEffectPhase);
+
+      expect(leadPokemon.getTag(BattlerTagType.JAW_LOCK)).toBeDefined();
+      expect(enemyPokemon.getTag(BattlerTagType.JAW_LOCK)).toBeDefined();
```
Comment (torranx) [inline src/test/moves/jaw_lock.test.ts:104]: why is the tag defined here on `MoveEffectPhase` but on other tests, its not?

Comment (innerthunder) [reply src/test/moves/jaw_lock.test.ts:104]: the other tests use
```ts
await game.phaseInterceptor.to(MoveEffectPhase, false)
```
so they check for the tag before `MoveEffectPhase` gets to run.

**Review by torranx — APPROVED** [review]
code's lgtm (did not test)

File: src/test/moves/jaw_lock.test.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+      await game.phaseInterceptor.to(FaintPhase);
+
+      expect(leadPokemon.getTag(BattlerTagType.JAW_LOCK)).toBeUndefined();
+      expect(enemyPokemon.getTag(BattlerTagType.JAW_LOCK)).toBeUndefined();
+    }, TIMEOUT
+  );
+
+  it(
+    "should only trap the user until the target faints",
```
Comment (Tempo-anon) [inline src/test/moves/jaw_lock.test.ts:92]: Shouldn't there be an `expect` in here to see the user and target get the jaw lock tag when the move is used but before either of the target is KO'd?

File: src/data/move.ts [hunk 4]
```diff
@@ -4390,6 +4390,30 @@ export class GulpMissileTagAttr extends MoveEffectAttr {
   }
 }
 
+/**
+ * Attribute used for moves that add a tag to the user and target, such as Jaw Lock
+ * @extends AddBattlerTagAttr
+ */
+export class AddLinkedBattlerTagAttr extends AddBattlerTagAttr {
```
Comment (Tempo-anon) [inline src/data/move.ts:4397]: Is this tag ever removed? I don't think I see that part

File: src/test/moves/jaw_lock.test.ts
Comment (Tempo-anon) [inline src/test/moves/jaw_lock.test.ts:1]: nit: Should probably also include some edge case tests for like double battles since if your Pokemon jaw locks opponent A, then jaw locks opponent B, then KOs opponent A they should still be trapped until opponent B is also KO'd

File: src/data/move.ts (hunk 4 above)
Comment (innerthunder) [reply src/data/move.ts:4397]: Only on faint (where `lapseTags(BattlerTagType.FAINT)` is called) and whenever `Pokemon.resetSummonData()` is called

File: src/test/moves/jaw_lock.test.ts
Comment (innerthunder) [reply src/test/moves/jaw_lock.test.ts:1]: [Bulba](https://bulbapedia.bulbagarden.net/wiki/Jaw_Lock_(move)) says the trap effect shouldn't apply to opponent B at all in this case, since the user is already trapped.

Seems like I need to add some logic to account for overlapping effects either way, though.

**Review by frutescens — APPROVED** [review]

File: src/data/move.ts [hunk 5]
```diff
@@ -4403,6 +4403,35 @@ export class GulpMissileTagAttr extends MoveEffectAttr {
...
```
Comment (f-fsantos) [inline src/data/move.ts:4410]: Should `Destiny Bond` also apply to this?

Comment (frutescens) [reply src/data/move.ts:4410]: Would moves like octolock, thousand waves, spirit shackle, and anchor shot also benefit from this attr as well?

Comment (frutescens) [reply src/data/move.ts:4410]: Possible Reference: 
https://bulbapedia.bulbagarden.net/wiki/Escape_prevention
Future Sight and Doom Desire too maybe?

Comment (innerthunder) [reply src/data/move.ts:4410]: As discussed earlier in the Discord, this only applies to Jaw Lock

**Review by flx-sta — APPROVED** [review]

**Review by xsn34kzx — APPROVED** [review]

**Comment by Snailman11** [comment]
Link to Discord's (Jaw Lock Still Prevents User Switch Out after making opponent faint)
https://discord.com/channels/1125469663833370665/1271997915556089936
